modules/hawkular-wildfly-agent-download/src/main/java/org/hawkular/component/wildflymonitor/WildFlyAgentServlet.java

```diff
@@ -86,37 +103,122 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
-            resp.setContentLength((int) agentJar.length());
-            resp.setDateHeader("Last-Modified", agentJar.lastModified());
-            try (FileInputStream agentJarStream = new FileInputStream(agentJar)) {
-                copy(agentJarStream, resp.getOutputStream());
+            resp.setHeader("Content-Disposition", "attachment; filename=" + agentInstallerJar.getName());
+            resp.setContentLength((int) agentInstallerJar.length());
+            resp.setDateHeader("Last-Modified", agentInstallerJar.lastModified());
+
+            ZipFile agentInstallerZip = new ZipFile(agentInstallerJar);
+            ZipOutputStream zos = new ZipOutputStream(resp.getOutputStream(), StandardCharsets.UTF_8);
+            final String wildflyHome = getValueFromQueryParam(req, AGENT_INSTALLER_PROPERTY_WILDFLY_HOME, null);
+            final String hawkularServer = getValueFromQueryParam(req, AGENT_INSTALLER_PROPERTY_HAWKULAR_SERVER,
+                    "http://localhost:8080");
+            String defaultModuleDist = hawkularServer.startsWith("http:") ? hawkularServer +
+                    "/hawkular/wildfly-agent/download" : "http://localhost:8080/wildfly-agent/download";
+            final String moduleDist = getValueFromQueryParam(req, AGENT_INSTALLER_PROPERTY_MODULE_DIST,
+                    defaultModuleDist);
+            final String username = getValueFromQueryParam(req, AGENT_INSTALLER_PROPERTY_USERNAME, null);
+            final String password = getValueFromQueryParam(req, AGENT_INSTALLER_PROPERTY_PASSWORD, null);
+
+
+            for (Enumeration<? extends ZipEntry> e = agentInstallerZip.entries(); e.hasMoreElements(); ) {
+                ZipEntry entryIn = e.nextElement();
+                if (!entryIn.getName().equalsIgnoreCase(AGENT_INSTALLER_PROPERTIES_FILE_NAME)) {
+                    // skip everything else
+                    zos.putNextEntry(entryIn);
+                    try (InputStream is = agentInstallerZip.getInputStream(entryIn)) {
+                        byte[] buf = new byte[4096];
+                        int len;
+                        while ((len = (is.read(buf))) > 0) {
+                            zos.write(buf, 0, len);
+                        }
+                    }
+                } else {
+                    zos.putNextEntry(new ZipEntry(AGENT_INSTALLER_PROPERTIES_FILE_NAME));
+                    try (BufferedReader br =
+                                 new BufferedReader(
+                                         new InputStreamReader(agentInstallerZip.getInputStream(entryIn),
+                                                 StandardCharsets.UTF_8)
+                                 )) {
+                        String line;
+                        while ((line = br.readLine()) != null) {
+                            //todo: it'd be better to reuse the strings from the agent installer module to keep it DRY
+                            // and resilient against the changes
```

Comment:
I have code somewhere that can do this kind of thing. I'll dig it out and we can use it here. It doesn't replace commented variables, but it does retain comments in the file. That would at least make this less dependent on things like commented example values of "/opt/wildfly" in the .properties file.